### ISSUE TITLE
Some Questions about Training and Validation.

### ISSUE THREAD

**Issue by BualaTravel** [VisualComputingInstitute/TrackR-CNN#43]
Thank you all so much for sharing this project. I would really appreciate your help.

I used TrackR-CNN with the pre-trained model “trackrcnn_init” to test on videos from overhead-located cameras recording people. However, on my overhead camera videos, the model detects few people. So I want to re-train it on ground truth sequences from my videos. Thus,  I have some questions to do this task.

1.	What software can I use in order to generate more ground truth images to annotate people and get these annotations in the .png format as you mention in the project’s link? 

2.	Would annotations in a .jpg format help with training speed due to the size of the jpg files?

3.	How many FPS and which ImageResolution is recommended to use before doing new annotated video frames to train TrackR-CNN?

4.	Once I get the annotated sequences, the paper mentions that MOTSChallenge was trained in a leaving-one-out fashion. So, as the paper mentions, instead, the KITTY MOTS was trained in a sort of 80% trained and 20%validation manner. Hence, which part of the code would allow me to switch between a leaving-one-out fashion and an 80% effort training way if I wanted to train a new dataset on whichever of two those ways? 

5.    After training during some predefined epochs, i.e., epochs=10, does the program update automatically the pre-trained-weight files after reaching this predefined amount let's say in a save_best_only way? This is because I am using Google Colab and I can just train it for 12 hours. (What would be your recommendation for doing this?)

6.	I am a little confused about the validation process. After training, shall I post-validate my new trained model using the mots_tools on a new validation dataset to see the results in sMOTSA MOTSA and MOTSP? Or how should it be?

7.	Which image resolution is recommended for doing training and validation? In the config file there are different resolutions for training and for validation.

8.	In Table-5 of the paper, it shows that  “GT Boxes (tight) + MG” has better results than “TrackR-CNN (ours)”. I do not know what the differences are between them. Please, could you elaborate on it a little bit more in detail?

**Comment by ahnonay** [comment]
Firstly, you can get the tracking IDs in the output files that TrackR-CNN generates, when you forward it on a dataset (see the "Forwarding" part of the Readme).

Regarding your numbered questions:

1. The annotations where partly hand made, partly generated by a segmentation network, as explained in the paper. You you write a small annotation tool to manually create these for your dataset.
2. I highly doubt this will affect training speed. In particular, we need the exact IDs stored in the png files, which we could not get from jpg since that uses lossy compression.
3. Higher image resolution will consume lots of resources during training. Higher FPS can often make tracking easier. As a reference, KITTI MOTS has 10 FPS and around 1200x400 image resolution.
4. You can change which sequences are trained/evaluated on using the "seq_ids_train" and "seq_ids_val" config flags (see e.g. KITTI_segtrack.py). For Kitty, we hardcoded the 80/20 split. For MOTSChallenge, we simply created four different configs, putting one sequence into seq_ids_val, respectively. (Arguably, that procedure is a bit ugly.)
5. The model is automatically saved to a new file (in the models/ subdirectory) each epoch (you can change how many past epochs are retained with the "max_saves_to_keep" config flag). You can then also restart the training from the last finished epoch by just running the same command again.
6. Yes, you can evaluate your model using mots_tools after training. We don't to evaluation during training to save time.
7. We use higher image resolution during validation, simply because we have more memory capacity during evaluation (don't need to store the forward pass feature maps). Using higher input resolution usually gives better performance.
8. In GT Boxes (tight) + MG, we take our annotated masks and compute enclosing bounding boxes from them. We then fill these bounding boxes with segmentation masks using the mask head from Mask R-CNN. In other words, this method gets ground truth bounding box input. It is like saying "How good can you get if you have perfect box tracking and then add masks?". It makes sense that the value here is higher than for TrackR-CNN (ours) (which does not get ground truth bounding box input).

**Comment by Crystian-001** [comment]
Hi Micheal Krause,

Thank you very much for your reply. I really appreciate it. Also,  how should I interpret this following info line?  I started the forwarding of detections of 440 png images at 17:29:30 and ended at 17:32:43. So roughly the process  lasted 3 min but I also have the oncoming info
 
Time for forwarding (s): 159.32030821300032 FPS for forwarding (wo. tracking): 2.7617320411642075

**Comment by BualaTravel** [comment]
Hi Micheal Krause,

Thank you very much for your reply. I really appreciate it. Also, how should I interpret this following info line? I started the forwarding of detections of 440 png images at 17:29:30 and ended at 17:32:43. So roughly the process lasted 3 min but I also have the oncoming info

Time for forwarding (s): 159.32030821300032 FPS for forwarding (wo. tracking): 2.7617320411642075

**Comment by pvoigtlaender** [comment]
The line says that the whole evaluation took 159.3 seconds (which is consistent with your ~3min). When neglecting the time for the tracking algorithm (i.e. just measuring the time used for the network, but the tracking algorithm should be almost free anyway), you get 2.76 frames per second speed.

**Comment by ahnonay** [comment]
Closing for now, please reopen if needed
